### PR TITLE
feat: customizable JSON-RPC error codes via new enum variant on `CallErrror`

### LINKED ISSUE #299: [server]: error types when register methods
Should we do something like `dyn Error` here instead? Or implement our own error trait that gives us the JSON-RPC error code along with the message?

_Originally posted by @maciejhirsz in https://github.com/paritytech/jsonrpsee/pull/295#r624936990_

### PR DESCRIPTION
**Pull request by niklasad1**
The motivation behind this change is that we need to provide functionality for users to configure their own error messages; that means error code, error message and data, https://www.jsonrpc.org/specification#error-object

Superseeds #393 

Closes #299

### CODE REVIEW

File: types/src/error.rs [hunk 1]
```diff
@@ -17,14 +18,24 @@ impl<T: fmt::Display> fmt::Display for Mismatch<T> {
 }
 
 /// Error that occurs when a call failed.
-#[derive(Debug, thiserror::Error)]
```
Comment (niklasad1) [inline types/src/error.rs:20]: `thiserror` didn't work with the introduced marker trait

File: types/src/traits.rs [hunk 2]
```diff
@@ -55,3 +55,20 @@ pub trait SubscriptionClient: Client {
...
+/// Trait for customizable [JSON-RPC error object](https://www.jsonrpc.org/specification#error-object)
+pub trait JsonRpcErrorT {
+	/// Error code (-32768 to -32000 are reserved and should not be used)
+	fn code(&self) -> i32;
+	/// Short description of the error.
+	// NOTE(niklasad1): this is a String because the underlying error may not be an constant
+	// and need to be included in the error.
+	fn message(&self) -> String;
+	/// A Primitive or Structured value that contains additional information about the error.
+	fn data(&self) -> Option<&serde_json::value::RawValue>;
```
Comment (niklasad1) [inline types/src/traits.rs:73]: might be annoying `serde_json::RawValue` here but at least we rexport it from types

**Comment by dvdplm** [comment]
cc @maciejhirsz Thoughts on this? 
For context: it seems like substrate only adheres to the spec (which mandates a full ["Error Object"](https://www.jsonrpc.org/specification#error_object)) some of the time? Possibly because it's verbose and annoying.

**Comment by niklasad1** [comment]
> cc @maciejhirsz Thoughts on this?
For context: it seems like substrate only adheres to the spec (which mandates a full "Error Object") some of the time? Possibly because it's verbose and annoying.

[example of this can be found here for context](https://github.com/paritytech/substrate/blob/master/client/rpc-api/src/author/error.rs#L106-L184) but I think these error codes is still quite useful in the context of substrate so we should really support it.

But I was thinking of two variants of `register_method`:

1. like old one, just require `StdError + Send + Sync` and just let jsonrpsee assign the error code
2. `register_method_with_custom_error`, require the error to impl `RpcErrorTrait`

File: http-server/src/tests.rs [hunk 3]
```diff
@@ -8,9 +8,36 @@ use jsonrpsee_test_utils::helpers::*;
...
+}
+
+impl JsonRpcErrorT for TestError {
+	fn code(&self) -> i32 {
+		match self {
+			Self::One => 1,
+			Self::Two => 2,
+		}
+	}
+	fn message(&self) -> String {
```
Comment (maciejhirsz) [inline http-server/src/tests.rs:30]: Instead of message, can't we just require `JsonRpcErrorT: Display` bound?

**Review by maciejhirsz — COMMENTED** [review]
This looks good to me, but I'm wondering if trait is the right approach. I think it might be better to have an enum like:

```rust
enum JsonRpcError {
    User(Box<dyn std::error::Error>),
    ParseError,
    InvalidError,
    // ...
}
```

With a `From<Box<dyn std::error::Error>>` impl, so we can auto cast all user errors into the `User` variant and give it some default error code, while also having the option to set codes for things like params parsing fail etc.

**Comment by niklasad1** [comment]
> With a `From<Box<dyn std::error::Error>>` impl, so we can auto cast all user errors into the `User` variant and give it some default error code, while also having the option to set codes for things like params parsing fail etc.

Right, that would be nice but the custom error messages won't work but if we could do:

```rust
enum JsonRpcError {
    // provide default error code
    User(Box<dyn std::error::Error>),
    // user provides error code and similar
    CustomUser { code: i32, message: String, data: JsonValue },
    // some specific error code
    ParseError,
    // some specific error code
    InvalidError,
    // ...
}
```

that would be quite flexible and better, what do you think @maciejhirsz? :)

**Comment by maciejhirsz** [comment]
Ye, that works I think.

File: types/src/error.rs [hunk 4]
```diff
@@ -19,12 +20,22 @@ impl<T: fmt::Display> fmt::Display for Mismatch<T> {
 /// Error that occurs when a call failed.
 #[derive(Debug, thiserror::Error)]
 pub enum CallError {
```
Comment (niklasad1) [inline types/src/error.rs:22]: we might want to put `non_exhaustive` on this but a little weird to just add some error code for non defined enum variant :P

File: types/src/error.rs [hunk 5]
```diff
@@ -19,12 +20,22 @@ impl<T: fmt::Display> fmt::Display for Mismatch<T> {
...
+	Failed(Box<dyn std::error::Error + Send + Sync>),
+	/// Custom error with specific JSON-RPC error code, message and data.
+	#[error("RPC Call failed: code: {code}, message: {message}, data: {data:?}")]
+	Custom {
+		/// JSON-RPC error code
+		code: i32,
+		/// Short description of the error.
+		message: String,
+		/// A primitive or structured value that contains additional information about the error.
+		data: Option<Box<RawValue>>,
```
Comment (niklasad1) [inline types/src/error.rs:37]: this might save allocations of the value is already a String

**Review by dvdplm — APPROVED** [review]
This should work.

**Review by maciejhirsz — APPROVED** [review]

**Pull request merged**